File: XitTests/XTFileChangesDataSourceTest.m
#import "XTTest.h"
#include "XTQueueUtils.h"
#import "Xit-Swift.h"

@interface XTFileChangesDataSourceTest : XTTest

@end

@implementation XTFileChangesDataSourceTest

- (void)testInitialCommit
{
  XTWindowController *winController = [[XTWindowController alloc] init];
  XTFileChangesDataSource *dataSource = [[XTFileChangesDataSource alloc] init];
  NSOutlineView *outlineView = [[NSOutlineView alloc] init];

  winController.selectedModel = [[XTCommitChanges alloc]
      initWithRepository:self.repository sha:self.repository.headSHA];
  dataSource.repository = self.repository;
  dataSource.repoController = (XTWindowController*)winController;
  outlineView.dataSource = dataSource;
  [dataSource reload];
  [self waitForRepoQueue];
  WaitForQueue(dispatch_get_main_queue());

  XCTAssertEqual(
      [dataSource outlineView:outlineView numberOfChildrenOfItem:nil], 1L);

  id item1 = [dataSource outlineView:outlineView child:0 ofItem:nil];

  XCTAssertEqualObjects([dataSource pathFor:item1], @"file1.txt");
  XCTAssertFalse([dataSource outlineView:outlineView isItemExpandable:item1]);
  XCTAssertEqual([dataSource changeFor:item1], XitChangeAdded);
}

@end

File: XitTests/XTFileChangesDataSourceTest.swift
import Foundation
import XCTest
@testable import Xit

class XTFileChangesDataSourceTest: XTTest
{
  func testInitialCommit() throws
  {
    let repoUIController = FakeRepoUIController(repository: repository)
    let dataSource = FileChangesDataSource(useWorkspaceList: false)
    let outlineView = NSOutlineView()
    let headSHA = try XCTUnwrap(repository.headSHA)
    let headCommit = try XCTUnwrap(GitCommit(sha: headSHA, repository: repository.gitRepo))
    
    repoUIController.repoController = GitRepositoryController(repository: repository)
    repoUIController.selection = CommitSelection(repository: repository,
                                                 commit: headCommit)
    objc_sync_enter(dataSource)
    dataSource.repoUIController = repoUIController
    objc_sync_exit(dataSource)
    outlineView.dataSource = dataSource
    dataSource.reload()
    waitForRepoQueue()
    WaitForQueue(DispatchQueue.main)
    
    XCTAssertEqual(dataSource.outlineView(outlineView,
                                          numberOfChildrenOfItem: nil),
                   1)
    
    let item1 = dataSource.outlineView(outlineView, child: 0, ofItem: nil)
    
    XCTAssertEqual(dataSource.path(for: item1), TestFileName.file1.rawValue)
    XCTAssertFalse(dataSource.outlineView(outlineView, isItemExpandable: item1))
    XCTAssertEqual(dataSource.change(for: item1), DeltaStatus.added)
  }
}
